src/storage/src/hummock/local_version/recent_versions.rs

```diff
@@ -0,0 +1,321 @@
+        for (table_id, info) in new_version.version().state_table_info.info() {
+            if let Some(prev_info) = self
+                .latest_version
+                .version()
+                .state_table_info
+                .info()
+                .get(table_id)
+            {
+                match info.committed_epoch.cmp(&prev_info.committed_epoch) {
+                    Ordering::Less => {
+                        unreachable!(
+                            "table {} has regress committed epoch {}, prev committed epoch {}",
+                            table_id, info.committed_epoch, prev_info.committed_epoch
+                        );
+                    }
+                    Ordering::Equal => {}
+                    Ordering::Greater => {
+                        has_table_committed = true;
+                    }
+                }
+            } else {
+                has_table_committed = true;
+            }
+        }
+        has_table_committed
+    }
+
+    #[must_use]
+    pub fn with_new_version(&self, version: PinnedVersion) -> Self {
+        assert!(version.version().id > self.latest_version.version().id);
+        let is_committed = self.has_table_committed(&version);
+        let recent_versions = if self.is_latest_committed {
+            let prev_recent_versions = if self.recent_versions.len() >= self.max_version_num {
+                assert_eq!(self.recent_versions.len(), self.max_version_num);
+                &self.recent_versions[1..]
+            } else {
+                &self.recent_versions[..]
+            };
+            let mut recent_versions = Vec::with_capacity(prev_recent_versions.len() + 1);
+            recent_versions.extend(prev_recent_versions.iter().cloned());
+            recent_versions.push(self.latest_version.clone());
+            recent_versions
+        } else {
+            self.recent_versions.clone()
+        };
+        Self {
+            latest_version: version,
+            is_latest_committed: is_committed,
+            recent_versions,
+            max_version_num: self.max_version_num,
+        }
+    }
+
+    pub fn latest_version(&self) -> &PinnedVersion {
+        &self.latest_version
+    }
+
+    /// Return the latest version that is safe to read `epoch` on `table_id`.
+    ///
+    /// `safe to read` means that the `committed_epoch` of the `table_id` in the version won't be greater than the given `epoch`
+    pub fn get_safe_version(
+        &self,
+        table_id: TableId,
+        epoch: HummockEpoch,
+    ) -> Option<PinnedVersion> {
+        if let Some(info) = self
+            .latest_version
+            .version()
+            .state_table_info
+            .info()
+            .get(&table_id)
+        {
+            if info.committed_epoch <= epoch {
+                Some(self.latest_version.clone())
+            } else {
+                self.get_safe_version_from_recent(table_id, epoch)
+            }
+        } else {
+            None
```

Comment:
Does it mean that for dropped table we always go through time travel even if it exists in one of the recent version?

Reply:
Yes. It is unexpected to query a dropped table, and therefore this is rarely hit. 

The reason for the current logic is to ensure that when querying the list of recent versions, the committed epoch of the table id is sorted so that we can easily use binary search. When the table id does not exist in a version, the table is either not created yet or dropped. When comparing a version with a specified query epoch, if the table is not created, the version is less than the query epoch, but if the table has been dropped, the version is greater than the query epoch, which means we cannot know the order between a query epoch and a version without the table_id.  Therefore, if we can ensure that when query the list of recent versions the table must exist in the latest version, the list of recent versions is sorted.